### PR TITLE
Add configurable auto-reconnect

### PR DESCRIPTION
**Pull request by DaelonSuzuka**
This addresses issue #156 by automatically attempting to reconnect to the language server.

This adds 3 new settings: `reconnect_automatically`, `reconnect_cooldown`, and `reconnect_attempts`.

The default values are to attempt to reconnect once per second, for 10 seconds, after which it displays the existing "Couldn't connect" error message.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -113,6 +113,21 @@
...
 					"description": "The scene file to run"
+				},
+				"godot_tools.reconnect_automatically": {
+					"type": "boolean",
+					"default": true,
+					"description": "Whether the plugin should attempt to reconnect"
+				},
+				"godot_tools.reconnect_cooldown": {
+					"type": "number",
+					"default": 1000,
```
Comment (Calinou) [inline package.json:124]: This seems pretty frequent, if I'm understanding it right. With 10 retries, it means the editor will only try to reconnect for 10 seconds (or perhaps 11).

Maybe this should be set to 3000 ms by default.

**Comment by Calinou** [comment]
Could you split the syntax file additions and changes to another pull request? This would make PRs easier to review individually.

**Comment by DaelonSuzuka** [comment]
The syntax stuff is 100% not supposed to be here, that's a mistake on my part. I'll fix this as soon as I can.

File: src/godot-tools.ts [hunk 2]
```diff
@@ -151,31 +151,61 @@ export class GodotTools {
 				this.connection_status.tooltip = `Connecting to the GDScript language server...`;
 				break;
 			case ClientStatus.CONNECTED:
+				this.retry = false
```
Comment (Calinou) [inline src/godot-tools.ts:154]:
```suggestion
				this.retry = false;
```

File: src/godot-tools.ts [hunk 3]
```diff
@@ -151,31 +151,61 @@ export class GodotTools {
...
+		}
+
+		this.retry = false
+		this.connection_status.text = `$(x) Disconnected`;
+		this.connection_status.tooltip = `Disconnected from the GDScript language server.`;
+
+		const message = `Couldn't connect to the GDScript language server.`;
+		vscode.window.showErrorMessage(message, 'Open Godot Editor', 'Retry', 'Ignore').then(item=>{
 			if (item == 'Retry') {
+				this.reconnection_attempts = 0
```
Comment (Calinou) [inline src/godot-tools.ts:202]:
```suggestion
				this.reconnection_attempts = 0;
```

File: src/godot-tools.ts [hunk 4]
```diff
@@ -151,31 +151,61 @@ export class GodotTools {
...
 				this.client.connect_to_server();
 			} else if (item == 'Open Godot Editor') {
 				this.client.status = ClientStatus.PENDING;
 				this.open_workspace_with_editor("-e").then(()=>{
 					setTimeout(()=>{
+						this.reconnection_attempts = 0
```
Comment (Calinou) [inline src/godot-tools.ts:208]:
```suggestion
						this.reconnection_attempts = 0;
```

File: src/godot-tools.ts (hunk 2 above)
Comment (Calinou) [inline src/godot-tools.ts]:
```suggestion
				this.retry = false;
```

File: src/godot-tools.ts [hunk 5]
```diff
@@ -35,19 +39,17 @@ export class GodotTools {
 		this.connection_status.text = "$(sync) Initializing";
 		this.connection_status.command = "godot-tool.check_status";
 		this.connection_status.show();
+		
+		this.reconnection_attempts = 0
```
Comment (Calinou) [inline src/godot-tools.ts:43]:
```suggestion
		this.reconnection_attempts = 0;
```

File: src/godot-tools.ts [hunk 6]
```diff
@@ -20,6 +20,10 @@ export class GodotTools {
 		this.client = new GDScriptLanguageClient(p_context);
 		this.client.watch_status(this.on_client_status_changed.bind(this));
 		this.connection_status = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right);
+
+		setInterval(()=> {
```
Comment (Calinou) [inline src/godot-tools.ts:24]:
```suggestion
		setInterval(() => {
```

File: src/godot-tools.ts [hunk 7]
```diff
@@ -151,31 +151,61 @@ export class GodotTools {
...
+
+		const message = `Couldn't connect to the GDScript language server.`;
+		vscode.window.showErrorMessage(message, 'Open Godot Editor', 'Retry', 'Ignore').then(item=>{
 			if (item == 'Retry') {
+				this.reconnection_attempts = 0
 				this.client.connect_to_server();
 			} else if (item == 'Open Godot Editor') {
 				this.client.status = ClientStatus.PENDING;
 				this.open_workspace_with_editor("-e").then(()=>{
 					setTimeout(()=>{
```
Comment (Calinou) [inline src/godot-tools.ts:207]:
```suggestion
					setTimeout(() =>{
```

File: src/godot-tools.ts [hunk 8]
```diff
@@ -151,31 +151,61 @@ export class GodotTools {
...
+		const message = `Couldn't connect to the GDScript language server.`;
+		vscode.window.showErrorMessage(message, 'Open Godot Editor', 'Retry', 'Ignore').then(item=>{
 			if (item == 'Retry') {
+				this.reconnection_attempts = 0;
 				this.client.connect_to_server();
 			} else if (item == 'Open Godot Editor') {
 				this.client.status = ClientStatus.PENDING;
 				this.open_workspace_with_editor("-e").then(()=>{
-					setTimeout(()=>{
+					setTimeout(() =>{
```
Comment (Calinou) [inline src/godot-tools.ts:207]:
```suggestion
					setTimeout(() => {
```

File: src/godot-tools.ts [hunk 9]
```diff
@@ -20,6 +20,10 @@ export class GodotTools {
 		this.client = new GDScriptLanguageClient(p_context);
 		this.client.watch_status(this.on_client_status_changed.bind(this));
 		this.connection_status = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right);
+
+		setInterval(() => {
+			this.retry_callback()
```
Comment (Calinou) [inline src/godot-tools.ts:25]:
```suggestion
			this.retry_callback();
```

**Comment by Calinou** [comment]
@DaelonSuzuka Could you look into rebasing this following the merge of https://github.com/godotengine/godot-vscode-plugin/pull/297? I'll then be able to merge this.

**Comment by DaelonSuzuka** [comment]
@Calinou I fetched the upstream, apparently. I have to admit this is deeper into git than I usually go. 

I think I made a made a mistake when I wrote this feature in my fork's master instead of making a branch for it.

**Comment by DaelonSuzuka** [comment]
I have no choice but to close this PR so I can rebuild my fork.

**Pull request closed without merging**